File: app/api/v1/endpoints/services.py
"""
Services API

Servis yönetimi için API endpoint'leri.
"""

from typing import List, Dict, Any, Optional
from fastapi import APIRouter, HTTPException, Depends, Body, Path, BackgroundTasks, Query, WebSocket, WebSocketDisconnect
from sqlalchemy.ext.asyncio import AsyncSession
import asyncio
import json
from datetime import datetime

from app.core.logger import setup_logging, get_logger
from app.services.service_wrapper import ServiceWrapper
from app.api.v1.schemas.service import (
    ServiceResponse, 
    ServiceStartRequest, 
    ServiceStopRequest,
    ServiceStatusResponse
)
from app.db.session import get_session
from app.services.service_manager import get_service_manager
from app.services.service_monitor import get_service_monitor
from app.api.deps import get_current_user

router = APIRouter(
    tags=["services"],
    # Geçici olarak auth middleware'i devre dışı bırak
    # dependencies=[Depends(get_current_user)]
)

logger = get_logger(__name__)

def get_service_wrapper() -> ServiceWrapper:
    """ServiceWrapper instance'ı döndürür."""
    return ServiceWrapper()

@router.get("/health", response_model=Dict[str, Any])
async def get_services_health(
    db: AsyncSession = Depends(get_session)
):
    """
    Tüm servislerin sağlık durumunu döndürür.
    """
    try:
        try:
            # Gerçek servisleri getirmeyi dene
            service_monitor = await get_service_monitor(db=db)
            if service_monitor is None:
                raise ValueError("Service monitor is None")
                
            services_health = await service_monitor.get_all_services_health()
            all_healthy = all(s.get("healthy", False) for s in services_health.values())
            
            return {
                "status": "ok" if all_healthy else "warning",
                "all_healthy": all_healthy,
                "services": services_health
            }
        except Exception as inner_e:
            logger.warning(f"Gerçek servis sağlık durumları alınamadı: {inner_e}")
            # Mock veri döndür
            mock_services = {
                "telegram_client": {
                    "status": "running",
                    "running": True,
                    "healthy": True,
                    "uptime": 1200,
                    "last_error": None
                },
                "database": {
                    "status": "running",
                    "running": True,
                    "healthy": True,
                    "uptime": 3600,
                    "last_error": None
                },
                "api_server": {
                    "status": "warning",
                    "running": True,
                    "healthy": False,
                    "uptime": 600,
                    "last_error": "Bağlantı sorunu"
                },
                "message_handler": {
                    "status": "stopped",
                    "running": False,
                    "healthy": False,
                    "uptime": 0,
                    "last_error": "Servis çalışmıyor"
                }
            }
            
            return {
                "status": "warning",
                "all_healthy": False,
                "services": mock_services
            }
    except Exception as e:
        logger.error(f"Servis sağlık durumları alınamadı: {e}")
        # Hata durumunda varsayılan sağlık durumu döndür
        return {
            "status": "error",
            "all_healthy": False,
            "services": {},
            "error": str(e)
        }

@router.get("/", response_model=Dict[str, Any])
async def list_all_services(
    db: AsyncSession = Depends(get_session)
):
    """
    Tüm servislerin durumunu döndürür.
    """
    try:
        # Gerçek servisleri getirmeyi dene
        try:
            service_manager = await get_service_manager(db=db)
            if service_manager is None:
                raise ValueError("Service manager is None")
                
            status = await service_manager.get_all_services_status()
            return {
                "success": True, 
                "services": status,
                "total": len(status),
                "active": sum(1 for s in status.values() if s.get("running", False))
            }
        except Exception as inner_e:
            logger.warning(f"Gerçek servis listesi alınamadı, mock veri kullanılıyor: {inner_e}")
            # Mock veri döndür
            mock_status = {
                "telegram_client": {"running": True, "healthy": True, "uptime": 1200},
                "database": {"running": True, "healthy": True, "uptime": 3600},
                "api_server": {"running": True, "healthy": False, "uptime": 600},
                "message_handler": {"running": False, "healthy": False, "uptime": 0}
            }
            
            return {
                "success": True, 
                "services": mock_status,
                "total": len(mock_status),
                "active": sum(1 for s in mock_status.values() if s.get("running", False))
            }
    except Exception as e:
        logger.error(f"Servis listesi alınamadı: {e}")
        raise HTTPException(status_code=500, detail=f"Servis listesi alınamadı: {str(e)}")

@router.get("/{service_name}", response_model=Dict[str, Any])
async def get_service_details(
    service_name: str = Path(..., description="Servis adı"),
    db: AsyncSession = Depends(get_session)
):
    """
    Belirli bir servisin detaylarını döndürür.
    """
    try:
        try:
            service_monitor = await get_service_monitor(db=db)
            if service_monitor is None:
                raise ValueError("Service monitor is None")
                
            details = await service_monitor.get_service_details(service_name)
            
            if "error" in details:
                raise HTTPException(status_code=404, detail=details["error"])
            
            return {"success": True, "service": details}
        except HTTPException:
            raise
        except Exception as inner_e:
            logger.warning(f"Gerçek servis detayları alınamadı ({service_name}), mock veri kullanılıyor: {inner_e}")
            
            # Mock veri döndür
            mock_services = {
                "telegram_client": {
                    "running": True,
                    "healthy": True,
                    "uptime": 1200,
                    "name": "telegram_client",
                    "status": "running",
                    "last_error": None,
                    "depends_on": ["database"]
                },
                "database": {
                    "running": True,
                    "healthy": True,
                    "uptime": 3600,
                    "name": "database",
                    "status": "running",
                    "last_error": None,
                    "depends_on": []
                },
                "api_server": {
                    "running": True,
                    "healthy": False,
                    "uptime": 600,
                    "name": "api_server",
                    "status": "warning",
                    "last_error": "Bağlantı sorunu",
                    "depends_on": ["database"]
                },
                "message_handler": {
                    "running": False,
                    "healthy": False,
                    "uptime": 0,
                    "name": "message_handler",
                    "status": "stopped",
                    "last_error": "Servis çalışmıyor",
                    "depends_on": ["telegram_client"]
                }
            }
            
            if service_name not in mock_services:
                raise HTTPException(status_code=404, detail=f"{service_name} servisi bulunamadı.")
                
            return {"success": True, "service": mock_services[service_name]}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Servis detayları alınamadı ({service_name}): {e}")
        raise HTTPException(status_code=500, detail=f"Servis detayları alınamadı: {str(e)}")

@router.post("/{service_name}/restart", response_model=Dict[str, Any])
async def restart_service(
    service_name: str = Path(..., description="Servis adı"),
    background_tasks: BackgroundTasks = BackgroundTasks(),
    db: AsyncSession = Depends(get_session)
):
    """
    Belirli bir servisi yeniden başlatır.
    """
    try:
        # İşlemi arka planda gerçekleştir (servis yeniden başlatma birkaç saniye sürebilir)
        service_monitor = await get_service_monitor(db=db)
        
        async def _restart_task():
            try:
                await service_monitor.restart_service(service_name)
            except Exception as e:
                logger.error(f"{service_name} servisi yeniden başlatılamadı: {e}")
        
        background_tasks.add_task(_restart_task)
        
        return {
            "success": True,
            "message": f"{service_name} servisi yeniden başlatılıyor."
        }
    except Exception as e:
        logger.error(f"Servis yeniden başlatılamadı ({service_name}): {e}")
        raise HTTPException(status_code=500, detail=f"Servis yeniden başlatılamadı: {str(e)}")

@router.post("/{service_name}/start", response_model=Dict[str, Any])
async def start_service(
    service_name: str = Path(..., description="Servis adı"),
    background_tasks: BackgroundTasks = BackgroundTasks(),
    db: AsyncSession = Depends(get_session)
):
    """
    Belirli bir servisi başlatır.
    """
    try:
        # İşlemi arka planda gerçekleştir
        service_manager = await get_service_manager(db=db)
        
        async def _start_task():
            try:
                await service_manager.restart_service(service_name)
            except Exception as e:
                logger.error(f"{service_name} servisi başlatılamadı: {e}")
        
        background_tasks.add_task(_start_task)
        
        return {
            "success": True,
            "message": f"{service_name} servisi başlatılıyor."
        }
    except Exception as e:
        logger.error(f"Servis başlatılamadı ({service_name}): {e}")
        raise HTTPException(status_code=500, detail=f"Servis başlatılamadı: {str(e)}")

@router.post("/{service_name}/stop", response_model=Dict[str, Any])
async def stop_service(
    service_name: str = Path(..., description="Servis adı"),
    background_tasks: BackgroundTasks = BackgroundTasks(),
    db: AsyncSession = Depends(get_session)
):
    """
    Belirli bir servisi durdurur.
    """
    try:
        service_manager = await get_service_manager(db=db)
        
        # Servis var mı kontrol et
        if service_name not in service_manager.services:
            raise HTTPException(status_code=404, detail=f"{service_name} servisi bulunamadı.")
        
        # İşlemi arka planda gerçekleştir
        async def _stop_task():
            try:
                service = service_manager.services[service_name]
                if hasattr(service, 'stop') and callable(service.stop):
                    await service.stop()
            except Exception as e:
                logger.error(f"{service_name} servisi durdurulamadı: {e}")
        
        background_tasks.add_task(_stop_task)
        
        return {
            "success": True,
            "message": f"{service_name} servisi durduruluyor."
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Servis durdurulamadı ({service_name}): {e}")
        raise HTTPException(status_code=500, detail=f"Servis durdurulamadı: {str(e)}")

@router.post("/{service_name}/config", response_model=Dict[str, Any])
async def update_service_config(
    service_name: str = Path(..., description="Servis adı"),
    config: Dict[str, Any] = Body(..., description="Servis konfigürasyonu"),
    db: AsyncSession = Depends(get_session)
):
    """
    Servis yapılandırmasını günceller.
    """
    try:
        service_monitor = await get_service_monitor(db=db)
        
        results = []
        for key, value in config.items():
            result = await service_monitor.update_service_config(service_name, key, value)
            results.append(result)
        
        # Tüm güncellemeler başarılı mı kontrol et
        all_success = all(r.get("success", False) for r in results)
        
        if all_success:
            return {
                "success": True,
                "message": f"{service_name} servisi için {len(config)} yapılandırma parametresi güncellendi."
            }
        else:
            # Başarısız güncelleme varsa, hata detaylarını döndür
            failed = [f"{k}: {results[i].get('error', 'Bilinmeyen hata')}" 
                     for i, k in enumerate(config.keys()) 
                     if not results[i].get("success", False)]
            
            return {
                "success": False,
                "message": f"Bazı yapılandırmalar güncellenemedi",
                "failed": failed
            }
    except Exception as e:
        logger.error(f"Servis yapılandırması güncellenemedi ({service_name}): {e}")
        raise HTTPException(status_code=500, detail=f"Servis yapılandırması güncellenemedi: {str(e)}")

@router.post("/start-all", response_model=Dict[str, Any])
async def start_all_services(
    background_tasks: BackgroundTasks = BackgroundTasks(),
    db: AsyncSession = Depends(get_session)
):
    """
    Tüm servisleri başlatır.
    """
    try:
        service_manager = await get_service_manager(db=db)
        
        # İşlemi arka planda gerçekleştir
        async def _start_all_task():
            try:
                await service_manager.start_services()
            except Exception as e:
                logger.error(f"Servisler başlatılamadı: {e}")
        
        background_tasks.add_task(_start_all_task)
        
        return {
            "success": True,
            "message": "Tüm servisler başlatılıyor."
        }
    except Exception as e:
        logger.error(f"Servisler başlatılamadı: {e}")
        raise HTTPException(status_code=500, detail=f"Servisler başlatılamadı: {str(e)}")

@router.post("/stop-all", response_model=Dict[str, Any])
async def stop_all_services(
    background_tasks: BackgroundTasks = BackgroundTasks(),
    db: AsyncSession = Depends(get_session)
):
    """
    Tüm servisleri durdurur.
    """
    try:
        service_manager = await get_service_manager(db=db)
        
        # İşlemi arka planda gerçekleştir
        async def _stop_all_task():
            try:
                await service_manager.stop_services()
            except Exception as e:
                logger.error(f"Servisler durdurulamadı: {e}")
        
        background_tasks.add_task(_stop_all_task)
        
        return {
            "success": True,
            "message": "Tüm servisler durduruluyor."
        }
    except Exception as e:
        logger.error(f"Servisler durdurulamadı: {e}")
        raise HTTPException(status_code=500, detail=f"Servisler durdurulamadı: {str(e)}")

@router.get("/status/summary", response_model=Dict[str, Any])
async def get_service_status_summary(
    db: AsyncSession = Depends(get_session)
):
    """
    Tüm servislerin özet durum bilgisini döndürür.
    """
    try:
        service_manager = await get_service_manager(db=db)
        status = await service_manager.get_all_services_status()
        
        total = len(status)
        active = sum(1 for s in status.values() if s.get("running", False))
        error = sum(1 for s in status.values() if s.get("error_count", 0) > 0)
        
        return {
            "success": True,
            "total": total,
            "active": active,
            "inactive": total - active,
            "error": error
        }
    except Exception as e:
        logger.error(f"Servis özeti alınamadı: {e}")
        raise HTTPException(status_code=500, detail=f"Servis özeti alınamadı: {str(e)}")

@router.websocket("/status-ws")
async def service_status_websocket(websocket: WebSocket):
    """
    WebSocket endpoint that streams service status information.
    """
    await websocket.accept()
    logger.info("WebSocket bağlantısı kabul edildi: servis durumu")
    
    try:
        # İlk durum bilgisini gönder
        db = await get_session()
        service_manager = await get_service_manager(db=db)
        initial_status = await service_manager.get_all_services_status()
        
        await websocket.send_json({
            "timestamp": datetime.utcnow().isoformat(),
            "services": initial_status
        })
        
        # Periyodik olarak güncellemeleri gönder
        while True:
            # Her 3 saniyede bir durum bilgisini güncelle
            await asyncio.sleep(3)
            
            try:
                db = await get_session()
                service_manager = await get_service_manager(db=db)
                status = await service_manager.get_all_services_status()
                
                await websocket.send_json({
                    "timestamp": datetime.utcnow().isoformat(),
                    "services": status
                })
            except Exception as e:
                logger.error(f"WebSocket servis durumu güncellenirken hata: {e}")
                await websocket.send_json({
                    "timestamp": datetime.utcnow().isoformat(),
                    "error": str(e)
                })
                
    except WebSocketDisconnect:
        logger.info("WebSocket bağlantısı kapatıldı: servis durumu")
    except Exception as e:
        logger.error(f"WebSocket hatası: {e}")
        try:
            await websocket.send_json({
                "timestamp": datetime.utcnow().isoformat(),
                "error": str(e)
            })
        except:
            pass 

File: app/frontend/admin-panel/src/api/services.ts
import axios from 'axios'

// API'nin çalıştığı adresi kullan
const API_URL = '/api/v1'
const USE_MOCK_DATA = false // Gerçek API hazır olmadığında geçici olarak mock verileri kullan

// Hata ayıklama için Axios interceptors ekle
axios.interceptors.request.use(config => {
  console.log(`${config.method?.toUpperCase()} isteği: ${config.url}`)
  return config
})

axios.interceptors.response.use(
  response => {
    console.log(`İstek başarılı: ${response.config.url}`)
    return response
  },
  error => {
    console.error(`İstek hatası: ${error.config?.url}`, error.response?.data || error.message)
    return Promise.reject(error)
  }
)

export type Service = {
  name: string
  status: string
  running: boolean
  healthy: boolean
  uptime: number
  last_error: string | null
  depends_on: string[]
}

export type ServiceStatusResponse = {
  status: string
  message: string
  details: Record<string, string>
}

export type HealthData = {
  status: string
  all_healthy: boolean
  services: Record<string, {
    status: string
    running: boolean
    healthy: boolean
    uptime: number
    last_error: string | null
  }>
}

// Mock veriler
const mockServices: Service[] = [
  {
    name: 'telegram_client',
    status: 'running',
    running: true,
    healthy: true,
    uptime: 1200,
    last_error: null,
    depends_on: ['database']
  },
  {
    name: 'database',
    status: 'running',
    running: true,
    healthy: true,
    uptime: 3600,
    last_error: null,
    depends_on: []
  },
  {
    name: 'api_server',
    status: 'warning',
    running: true,
    healthy: false,
    uptime: 600,
    last_error: 'Bağlantı sorunu',
    depends_on: ['database']
  },
  {
    name: 'message_handler',
    status: 'stopped',
    running: false,
    healthy: false,
    uptime: 0,
    last_error: 'Servis çalışmıyor',
    depends_on: ['telegram_client']
  }
]

const mockHealthData: HealthData = {
  status: 'warning',
  all_healthy: false,
  services: {
    telegram_client: {
      status: 'running',
      running: true,
      healthy: true,
      uptime: 1200,
      last_error: null
    },
    database: {
      status: 'running',
      running: true,
      healthy: true,
      uptime: 3600,
      last_error: null
    },
    api_server: {
      status: 'warning',
      running: true,
      healthy: false,
      uptime: 600,
      last_error: 'Bağlantı sorunu'
    },
    message_handler: {
      status: 'stopped',
      running: false,
      healthy: false,
      uptime: 0,
      last_error: 'Servis çalışmıyor'
    }
  }
}

// Tüm servisleri getir
export const getServices = async (): Promise<Service[]> => {
  if (USE_MOCK_DATA) {
    console.log('Mock veri kullanılıyor: getServices')
    return mockServices
  }

  try {
    const response = await axios.get<Service[]>(`${API_URL}/services`)
    return response.data
  } catch (error) {
    console.error('Servisler alınırken hata:', error)
    throw error
  }
}

// Belirli bir servisi getir
export const getService = async (serviceName: string): Promise<Service> => {
  if (USE_MOCK_DATA) {
    console.log(`Mock veri kullanılıyor: getService(${serviceName})`)
    const service = mockServices.find(s => s.name === serviceName)
    if (service) return service
    throw new Error('Servis bulunamadı')
  }

  try {
    const response = await axios.get<Service>(`${API_URL}/services/${serviceName}`)
    return response.data
  } catch (error) {
    console.error(`Servis alınırken hata (${serviceName}):`, error)
    throw error
  }
}

// Servisleri başlat
export const startService = async (serviceNames: string[]): Promise<ServiceStatusResponse> => {
  if (USE_MOCK_DATA) {
    console.log(`Mock veri kullanılıyor: startService(${serviceNames.join(', ')})`)
    return {
      status: 'success',
      message: 'Servisler başlatıldı',
      details: serviceNames.reduce((acc, name) => ({...acc, [name]: 'started'}), {})
    }
  }

  try {
    const response = await axios.post<ServiceStatusResponse>(`${API_URL}/services/start`, {
      services: serviceNames
    })
    return response.data
  } catch (error) {
    console.error('Servisler başlatılırken hata:', error)
    throw error
  }
}

// Servisleri durdur
export const stopService = async (serviceNames: string[], force = false): Promise<ServiceStatusResponse> => {
  if (USE_MOCK_DATA) {
    console.log(`Mock veri kullanılıyor: stopService(${serviceNames.join(', ')}, ${force})`)
    return {
      status: 'success',
      message: 'Servisler durduruldu',
      details: serviceNames.reduce((acc, name) => ({...acc, [name]: 'stopped'}), {})
    }
  }

  try {
    const response = await axios.post<ServiceStatusResponse>(`${API_URL}/services/stop`, {
      services: serviceNames,
      force
    })
    return response.data
  } catch (error) {
    console.error('Servisler durdurulurken hata:', error)
    throw error
  }
}

// Servis yeniden başlat
export const restartService = async (serviceName: string): Promise<ServiceStatusResponse> => {
  if (USE_MOCK_DATA) {
    console.log(`Mock veri kullanılıyor: restartService(${serviceName})`)
    return {
      status: 'success',
      message: `${serviceName} servisi yeniden başlatıldı`,
      details: {[serviceName]: 'restarted'}
    }
  }

  try {
    const response = await axios.post<ServiceStatusResponse>(`${API_URL}/services/${serviceName}/restart`)
    return response.data
  } catch (error) {
    console.error(`Servis yeniden başlatılırken hata (${serviceName}):`, error)
    throw error
  }
}

// Servis sağlık durumlarını getir
export const getServicesHealth = async (): Promise<HealthData> => {
  if (USE_MOCK_DATA) {
    console.log('Mock veri kullanılıyor: getServicesHealth')
    return mockHealthData
  }

  try {
    const response = await axios.get<HealthData>(`${API_URL}/services/health`)
    return response.data
  } catch (error) {
    console.error('Servis sağlık durumları alınırken hata:', error)
    throw error
  }
} 